### PR TITLE
GEOMESA-592,GEOMESA-697

### PR DESCRIPTION
**Pull request by dmcwhorter**
* Adding support for delimited ingest of Map attributes
* Fixing the encodings and decodings cache key in SimpleFeatureSerializer to return the correct encodings for different SFT's with the same field names

### CODE REVIEW

File: geomesa-feature/src/test/scala/org/locationtech/geomesa/feature/ScalaSimpleFeatureTest.scala [hunk 1]
```diff
@@ -79,9 +80,10 @@ class ScalaSimpleFeatureTest extends Specification {
       f.getProperties("a").head.getValue must not(throwA [org.opengis.feature.IllegalAttributeException])
 
       val prop = f.getProperty("a")
-      prop must not beNull;
-      prop.getName.getLocalPart mustEqual("a")
-      prop.getValue mustEqual(1)
+      prop should not beNull
+
+      prop.getName.getLocalPart mustEqual "a"
+      prop.getValue mustEqual 1
```
Comment (elahrvivaz) [inline geomesa-feature/src/test/scala/org/locationtech/geomesa/feature/ScalaSimpleFeatureTest.scala]: Did anything actually change in this file?

File: geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/SimpleFeatureTypes.scala [hunk 2]
```diff
@@ -122,6 +122,18 @@ object SimpleFeatureTypes {
     Option(ad.getUserData.get(OPT_CARDINALITY).asInstanceOf[String])
         .flatMap(c => Try(Cardinality.withName(c)).toOption).getOrElse(Cardinality.UNKNOWN)
 
+  def setMapTypes(ad: AttributeDescriptor, keyType: Class[_], valueType: Class[_]): Unit = {
+    ad.getUserData.put(USER_DATA_MAP_KEY_TYPE, keyType)
+    ad.getUserData.put(USER_DATA_MAP_VALUE_TYPE, valueType)
+  }
+
+  def getMapTypes(ad: AttributeDescriptor): Option[(Class[_], Class[_])] =
+    Option(ad.getUserData.get(USER_DATA_MAP_KEY_TYPE)).flatMap { keyClass =>
```
Comment (jnh5y) [inline geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/SimpleFeatureTypes.scala]: Can this be re-written as a for comprehension?

**Comment by elahrvivaz** [comment]
looks good.

File: geomesa-feature/src/test/scala/org/locationtech/geomesa/feature/ScalaSimpleFeatureTest.scala (hunk 1 above)
Comment (dmcwhorter) [inline geomesa-feature/src/test/scala/org/locationtech/geomesa/feature/ScalaSimpleFeatureTest.scala]: just formatting

File: geomesa-feature/src/test/scala/org/locationtech/geomesa/feature/kryo/KryoFeatureSerializerTest.scala [hunk 3]
```diff
@@ -107,7 +107,44 @@ class KryoFeatureSerializerTest extends Specification {
...
+        deserialized must not(beNull)
+        deserialized.getType mustEqual sf.getType
+        deserialized.getAttributes mustEqual sf.getAttributes
+      }
+    }
+
+    // NB: this doesn't actually seem to cause the error I was seeing, but
+    // ScaldingDelimitedIngestJobTest in geomesa-tools does cause it...
+    "correctly serialize with a type having the same field names but different field types" in {
+      val spec = "a:Integer,m:List[String],l:Map[Double,Date],dtg:Date,*geom:Point:srid=4326"
```
Comment (cne1x) [inline geomesa-feature/src/test/scala/org/locationtech/geomesa/feature/kryo/KryoFeatureSerializerTest.scala]: [picky]  Is there a reason the list attribute is called ```m``` while the map attribute is called ```l```?  Mind blown.

**Comment by jnh5y** [comment]
Committed as a5c215e2093208ec11bef2162a0d758ce1dad29a on 1.5.

**Pull request closed without merging**